Add Drag and Drop API

blandinw opened the pull request:
Fixes #56 

This PR implements the Drag and Drop API as described here: https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API

I tried to cover as much as possible with the tests and example.

A few things to note:
- the standard JS API is sometimes awkward to work with. Notably `DataTransferItem.getAsString` takes a callback, whereas `DataTransferItem.getAsFile` directly returns the file
- I decided to go with both a Future-based and callback-based version (not sure about `'static` lifetime in `get_as_string_callback`)
- I could not test `get_as_string()` and `get_as_string_callback()` without async tests (#128)
- another awkward thing is that the spec requires browsers to make some properties read-only except during some events firing. The issue is that we cannot simulate these events in a way that makes the datastore read/write (see tests in `drag.rs`). An example is that it's only possible to set `event.dataTransfer.effectAllowed` during `ondragstart`

blandinw commented:
Don't think the CI failure is due to my changes, please advise https://travis-ci.org/koute/stdweb/jobs/415782023

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,623 @@
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/getAsString)
+    #[cfg(feature = "futures-support")]
+    pub fn get_as_string( &self ) -> oneshot::Receiver<String> {
```
I'm not against having a future-y method, but it probably should be the other way around. (Have the callback version be called `get_as_string`, and the futures version something else, since the callback one is the original.)

roblabla commented:
I've been using this PR for some stuff, here are some thoughts:

- https://github.com/koute/stdweb/pull/258/files#diff-a5794c191f7b8ecca9844d2bd24791f5R508 is problematic. Some browser return non-standard kinds. Firefox does this, for instance, returning kind "other". You can test this by dropping an image.
- It'd be nice if DataTransferItemList implemented IntoIterator/Iterator, so it could be used in a for loop. FileList already does this.

blandinw commented:
@roblabla thank you for the feedback!
- I'll implement the IntoIterator/Iterator trait for DataTransferItemList
- I could not repro the item having kind "other" by dropping an image in Firefox. Are you able to repro in the examples/drag/ project from this PR? In any case, we should probably support a Nonstandard(String) enum member or similar. Thoughts?

roblabla commented:
I'm using Firefox Nightly, and I'm dropping an image from an <img> tag into a drop zone on the same page.

blandinw commented:
Fixed, thank you

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,697 @@
+            "move" => DropEffect::Move,
+            "link" => DropEffect::Link,
+            "none" => DropEffect::None,
+            other => panic!("Expected valid dropEffect value, got {:?}", other),
```
If it's true that this can return non-standard values then something like this would be good:

```
pub enum DropEffect {
    None,
    Copy,
    Move,
    Link,
    Other( OtherDropEffect )
}
impl DropEffect {
    fn as_str( &self ) -> &str { ... }
}
struct OtherDropEffect( String );
```

Basically prevent direct access to the string. But I'd really like if someone could investigate if this is really necessary.

koute replied:
It's more future proof in case we have to add more variants.

(I also forgot to mention that if we add the the `Other` variant then it should be `#[doc(hidden)]` and the whole enum should be explicitly marked as non-exhaustive in the docs.)

Let's assume we give the user direct access to the string, so they write the following code:

```
match drop_effect {
    DropEffect::Copy => { ... },
    DropEffect::Other( other ) if other == "foobar" => { ... },
    _ => { ... }
}
```

Now we decide to add `DropEffect::Foobar`. The user updates, and their code is broken. If we don't give the user a direct access to that variant then they have to write their code like this:

```
match drop_effect {
    DropEffect::Copy => { ... },
    other => if other.as_str() == "foobar" { ... },
    _ => { ... }
}
```

And now their code still works after we add `DropEffect::Foobar`.

blandinw replied:
Love it, thanks I'll implement this

blandinw replied:
@koute here are a few points regarding this, to make sure we are on the same page:
- we only needed an extra variant for the `DataTransferItemKind` enum only, not for the other two `DropEffect` and `EffectAllowed` (tested + googled around)
- seems like we'd want to use [#[non_exhaustive]](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md), but it is an experimental Rust feature at the moment. I added a TODO to use it in the future.
- I made the type `OtherKind` public, since variant cannot leak private types ([E0446](https://doc.rust-lang.org/error-index.html#E0446))
- I made the string field private so that the user must use `.as_str()`
- `.as_str()` has to be used on the variable of type `DataTransferItemKind`, I'm assuming that's what you meant in your example

I used it like so in the example project:
```
let kind_str = match x.kind() {
    DataTransferItemKind::String => String::from("a string"),
    DataTransferItemKind::File => String::from("a file"),
    DataTransferItemKind::Other(_) if x.kind().as_str() == "other" => String::from("an expected other kind"),
    DataTransferItemKind::Other(_) => format!("an unexpected other kind \"{}\"", x.kind().as_str()),
};
```

koute replied:
Sounds good to me! However, we shouldn't use the `DataTransferItemKind::Other` explicitly in the example - it's marked `#[doc(hidden)]` after all! The whole point of it is that the user should never ever access that variant directly in any way.

So using your example:

```
let kind_str = match x.kind() {
    DataTransferItemKind::String => String::from("a string"),
    DataTransferItemKind::File => String::from("a file"),
    item_kind if item_kind.kind().as_str() == "other" => String::from("an expected other kind"),
    item_kind => format!("an unexpected other kind \"{}\"", item_kind.kind().as_str()),
};
```

Actually, one more thing - let's rename the `Other` variant to, I don't know, `__Other` or something like that, to make it less likely that someone will use it by mistake.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,697 @@
+    /// Set the image to be used for dragging if a custom one is desired.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/setDragImage)
+    pub fn set_drag_image( &self, img: &ImageElement, x_offset: u32, y_offset: u32 ) {
```
The specs say that the offsets should be `i32`. (Side note: does it actually make sense to pass negative values here?)

blandinw replied:
> The values are expressed as distances in CSS pixels from the left side and from the top side of the image respectively
I don't think it makes sense to pass negative values here, but I'll update to match the spec.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,697 @@
+    /// An unsigned long that is the number of drag items in the list.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/length)
+    pub fn len( &self ) -> u64 {
```
This should return `u32` according to the specs.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,697 @@
+    /// a DataTransferItem object for the new item.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/add)
+    pub fn add( &self, item: DataTransferItemAdd ) -> Option<DataTransferItem> {
```
1. Get rid of the enum and split this into two methods - `add` and `add_file` perhaps?
2. This can throw `NotSupportedError` in case of the string variant, so it should return `Result< Option< DataTransferItem >, NotSupportedError >`.

blandinw replied:
1. I will split this into `add_string` and `add_file`

blandinw replied:
We could also replace the type `DataTransferItem` with two types `DataTransferItemString` and `DataTransferItemFile`, for better safety. I'll let you decide what the right trade-off is between convenience and spec accuracy.

koute replied:
It's generally simpler and more Rust-y to just have two methods instead of emulating overloading with enums, (And it's one less enum to `use`.) so please make it into two methods. (:

blandinw replied:
I did already (`add_string` and `add_file`), see the current version of the PR

blandinw replied:
What I meant, is that since we have two methods, we can have two return types encoding the two different behaviors. By default, I'll keep one return type.

koute replied:
Ah, right, sorry! I replied before looking at the code. (:

Hmm... yes, you're right, we could have two different return types, although it's probably not worth it in this case.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,697 @@
+    /// Removes the drag item from the list at the given index.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/remove)
+    pub fn remove( &self, index: u32 ) {
```
This can throw `InvalidStateError`

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,697 @@
+    /// (or null if the drag item is not a file)
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/getAsFile)
+    pub fn get_as_file( &self ) -> File {
```
This should return `Option` according to the specs.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,723 @@
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/add)
+    // https://www.w3.org/TR/html51/editing.html#ref-for-dom-datatransferitemlist-add-1
+    pub fn add_string( &self, data: String, ty: String ) -> Result<Option<DataTransferItem>, NotSupportedError> {
```
This should take `&str`s instead of `String`s.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,723 @@
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/add)
+    // https://www.w3.org/TR/html51/editing.html#ref-for-dom-datatransferitemlist-add-2
+    pub fn add_file( &self, file: File ) -> Option<DataTransferItem> {
```
Taking `&File` here should make more sense.

koute commented on src/webapi/events/drag.rs:
```diff
@@ -0,0 +1,723 @@
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Element/dropEffect)
+    // https://www.w3.org/TR/html51/editing.html#ref-for-dom-datatransfer-dropeffect-2
+    pub fn set_drop_effect( &self, value: DropEffect ) {
```
If we're going to add the other variant (which will contain a `String` internally) then this should be taken as `&DropEffect`. (Since the type won't be `Copy` anymore.)

blandinw replied:
See my comment in the other thread: after looking into it, the `Other` variant was only needed for `DataTransferItemKind` so I don't think this comment applies anymore

blandinw commented:
I pushed a new commit to address your comments. See the few points about the `Other` variants in the relevant discussion thread.

This is my first contribution to a Rust project and I'm grateful you took the time to teach me some idioms and help me improve the code!

Thank you for your patience, let's ship this 😊

blandinw commented:
I renamed `Other` to `__Other` and updated the example as such (have to take `kind` by ref instead of move since the next arm also needs it):
```
let kind_str = match x.kind() {
    DataTransferItemKind::String => String::from("a string"),
    DataTransferItemKind::File => String::from("a file"),
    ref kind if kind.as_str() == "other" => String::from("an expected other kind"),
    kind => format!("an unexpected other kind \"{}\"", kind.as_str()),
};
```

koute commented:
LGTM! Thanks!

I'll merge it and release it as soon as I'll clean up the current mess of compilation under Emscripten being broken.

spease commented:
Just saw this closed out my issue...this is awesome, thanks!
